onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -1794,50 +1832,44 @@
         for (auto it = node->OutputNodesBegin(); it != node->OutputNodesEnd(); ++it) {
           // if the output node is not in the same stream, generate a trigger point
           if (node_stream_map_[it->Index()] != i) {
-            node_to_trigger_points[node_index] = num_trigger_points;
-            num_trigger_points++;
+            diverge_stream_from_node[node_index] = num_notifications;
+            num_notifications++;
             break;
           }
         }
       }
     }
+
     for (size_t i = 0; i < num_logic_streams_; ++i) {
       for (auto node_index : stream_nodes_[i]) {
         auto* node = graph_viewer_.GetNode(node_index);
         for (auto it = node->OutputNodesBegin(); it != node->OutputNodesEnd(); ++it) {
-          bool requires_notification = false;
           // !! special case, Shape op's output is ready for all the EPs, so don't need notification
           if (node->OpType() != "Shape") {
             for (auto* output : node->OutputDefs()) {
               if (output->Exists()) {
                 if (std::find(it->InputDefs().begin(), it->InputDefs().end(), output) != it->InputDefs().end()) {
                   OrtValueIndex output_arg_idx;
                   ORT_THROW_IF_ERROR(ort_value_name_idx_map_.GetIdx(output->Name(), output_arg_idx));
-                  // there are two cases we need notification:
-                  // 1. the consumer is not in the same stream
-                  // 2. the consumer is in the same stream(non-cpu device), but it consumer a CPU tensor from an non-shape op.
-                  //    for example, a resize cuda kernel consumer a tensor from MemCpyToHost cuda kernel on the same stream.
-                  //    in this case, the FIFO can't gurantee the cpu tensor is ready when resize kernel is launching
-                  if (node_stream_map_[it->Index()] != i ||
-                      (node_stream_map_[it->Index()] == i &&
+                  // the consumer is in the same stream(non-cpu device), but it consumes a CPU tensor from an non-shape op.
+                  // for example, a resize cuda kernel consumer a tensor from MemCpyToHost cuda kernel on the same stream.
+                  // in this case, the FIFO can't guarantee the cpu tensor is ready when resize kernel is launching
+                  if ((node_stream_map_[it->Index()] == i &&
                        execution_plan[i]->device_.Type() != OrtDevice::CPU &&
                        plan_.allocation_plan[output_arg_idx].location.device.Type() == OrtDevice::CPU)) {
-                    requires_notification = true;
+                    node_to_notification[node_index] = num_notifications;
```

Comment:
> node_to_notification[node_index] = num_notifications; [](http://example.com/codeflow?start=19&length=54)

"node_to_notification[node_index] = num_notifications++;"?
And we are deliberately incrementing  num_notifications both on line 1836 and here?

Reply:
Yes. size(notification_owner) == size(diverge_stream_from_node) + size(node_to_notification). See the newly added comments, both diverge_stream_from_node and node_to_notification is capable of launching ActivateNotification step.